### PR TITLE
[ntuple] various interface improvements

### PR DESCRIPTION
**Pull request by jblomer**
Change raw pointer methods to shared pointers or const ref. Remove unsafe and unnecessary methods.

Add `RNTupleModel::GenerateBulk()`.

### CODE REVIEW

File: tree/ntuple/v7/inc/ROOT/RNTupleModel.hxx [hunk 1]
```diff
@@ -279,34 +280,11 @@ public:
...
-   }
-
    /// Adds a top-level field based on existing fields. The mapping function is called with the qualified field names
    /// of the provided field and the subfields.  It should return the qualified field names used as a mapping source.
    /// Projected fields can only be used for models used to write data.
    RResult<void> AddProjectedField(std::unique_ptr<Detail::RFieldBase> field,
                                    std::function<std::string(const std::string &)> mapping);
-
-   template <typename T>
-   T *Get(std::string_view fieldName) const
```
Comment (enirolf) [inline tree/ntuple/v7/inc/ROOT/RNTupleModel.hxx:304]: Would it be an idea to have a `Get<T>` (perhaps named differently) that returns a shared pointer to the object? The use case would be for type-erased fields. Right now the (only) way to do it is through `ntuple->GetModel().GetDefaultEntry()->GetPtr<T>("fieldname")`, which feels a bit unwieldy.

File: tree/ntuple/v7/inc/ROOT/RNTupleModel.hxx [hunk 2]
```diff
@@ -325,7 +303,10 @@ public:
    /// In a bare entry, all values point to nullptr. The resulting entry shall use BindValue() in order
    /// set memory addresses to be serialized / deserialized
    std::unique_ptr<REntry> CreateBareEntry() const;
-   REntry *GetDefaultEntry() const;
+   Detail::RFieldBase::RBulk GenerateBulk(std::string_view fieldName) const;
```
Comment (enirolf) [inline tree/ntuple/v7/inc/ROOT/RNTupleModel.hxx:306]: Perhaps add some documentation

**Review by enirolf — CHANGES_REQUESTED** [review]
Thanks! I have left two small comments.

File: tree/ntuple/v7/inc/ROOT/RNTupleModel.hxx (hunk 1 above)
Comment (jblomer) [reply tree/ntuple/v7/inc/ROOT/RNTupleModel.hxx:304]: That's a possibility, a getter acting as a shortcut for `GetModel().GetDefaultEntry()->GetPtr<T>`. On the other hand, callers can could also first get the entry and then call (multiple times) `entry.GetPtr<T>()`. How about we see if users stumble across it and defer the addition of the method until then?

Comment (enirolf) [reply tree/ntuple/v7/inc/ROOT/RNTupleModel.hxx:304]: Yes fair enough, sounds good!

**Review by enirolf — APPROVED** [review]
It looks good to me, but following from the discussion on shared pointers on Friday maybe someone else should also have a look.

**Comment by hahnjo** [comment]
I already had a quick look last week and didn't have any further comments (except maybe that `GenerateBulk` is first used in the next commit titled "return const ref from RNTupleReader::GetModel()" :see_no_evil:)

**Comment by smuzaffar** [comment]
@jblomer , we are trying to test latest root master changes in [CMSSW](https://github.com/cms-sw/cmsdist/pull/8973) and our build failed with errors like [a]. Looks like this change is causing cmssw to fail. I guess we need to update cmssw to use `model.GetDefaultEntry()->GetPtr<T>(name)` instead of `model.Get<T>(name)` ... right? What about `addField<T>()` calls?

We also want same cmssw code to work for ROOT 6.26 and above. So should we add some `#if ROOT_VERSION_CODE >= ROOT_VERSION(6,31,0)` blocks ?


[a] https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-6f60ec/37085/build-logs/PhysicsTools/NanoAOD/log.html
```
  /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_14_0_ROOT6_X_2024-01-28-2300/src/PhysicsTools/NanoAOD/plugins/rntuple/TriggerOutputFields.cc:126:38: error: 'class ROOT::Experimental::RNTupleModel' has no member named 'Get'
   126 |   const auto* existing_field = model.Get<bool>(name);
      |                                      ^~~
  /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_14_0_ROOT6_X_2024-01-28-2300/src/PhysicsTools/NanoAOD/plugins/rntuple/TriggerOutputFields.cc:126:42: error: expected primary-expression before 'bool'
   126 |   const auto* existing_field = model.Get<bool>(name);
      |                                          ^~~~
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_14_0_ROOT6_X_2024-01-28-2300/src/PhysicsTools/NanoAOD/plugins/rntuple/NanoAODRNTupleOutputModule.cc: In member function 'void NanoAODRNTupleOutputModule::CommonEventFields::createFields(ROOT::Experimental::RNTupleModel&)':
  /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_14_0_ROOT6_X_2024-01-28-2300/src/PhysicsTools/NanoAOD/plugins/rntuple/NanoAODRNTupleOutputModule.cc:77:28: error: expected primary-expression before '>' token
    77 |       model.AddField<UInt_t>("run", &m_run);
      |                            ^
  /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_14_0_ROOT6_X_2024-01-28-2300/src/PhysicsTools/NanoAOD/plugins/rntuple/NanoAODRNTupleOutputModule.cc:77:30: error: left operand of comma operator has no effect [-Werror=unused-value]
    77 |       model.AddField<UInt_t>("run", &m_run);
      |                              ^~~~~
  /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_14_0_ROOT6_X_2024-01-28-2300/src/PhysicsTools/NanoAOD/plugins/rntuple/NanoAODRNTupleOutputModule.cc:78:28: error: expected primary-expression before '>' token
    78 |       model.AddField<UInt_t>("luminosityBlock", &m_luminosityBlock);
      |                            ^
  /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_14_0_ROOT6_X_2024-01-28-2300/src/PhysicsTools/NanoAOD/plugins/rntuple/NanoAODRNTupleOutputModule.cc:78:30: error: left operand of comma operator has no effect [-Werror=unused-value]
    78 |       model.AddField<UInt_t>("luminosityBlock", &m_luminosityBlock);
      |                              ^~~~~~~~~~~~~~~~~
  /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_14_0_ROOT6_X_2024-01-28-2300/src/PhysicsTools/NanoAOD/plugins/rntuple/NanoAODRNTupleOutputModule.cc:79:35: error: expected primary-expression before '>' token
    79 |       model.AddField<std::uint64_t>("event", &m_event);
      |                                   ^
  /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_14_0_ROOT6_X_2024-01-28-2300/src/PhysicsTools/NanoAOD/plugins/rntuple/NanoAODRNTupleOutputModule.cc:79:37: error: left operand of comma operator has no effect [-Werror=unused-value]
    79 |       model.AddField<std::uint64_t>("event", &m_event);
      |                                     ^~~~~~~

```

**Comment by smuzaffar** [comment]
@jblomer @hahnjo  any suggestion about https://github.com/root-project/root/pull/14454#issuecomment-1914756475 ?

**Comment by hahnjo** [comment]
> I guess we need to update cmssw to use `model.GetDefaultEntry()->GetPtr<T>(name)` instead of `model.Get<T>(name)` ... right?

Yes, but with yesterday's changes I think it will be `model.GetDefaultEntry().GetPtr<T>(name)` because `GetDefaultEntry` returns a reference now. In general, there will be some more interface changes coming in the near future...

> What about `addField<T>()` calls?

That one is a bit more complicated; after a quick look, I think `CommonEventFields` should call `MakeField`, store the returned `std::shared_ptr` and use that one in `fill(const edm::EventID& id)`.

**Comment by smuzaffar** [comment]
thanks @hahnjo , I have opened https://github.com/cms-sw/cmssw/pull/43834

**Pull request merged**